### PR TITLE
feat(kit): `Number` allows to enter full width numbers

### LINKED ISSUE #865: 🚀 - Support for full width numbers for `maskitoNumberOptionsGenerator`
### Which package(s) are relevant/related to the feature request?

@maskito/kit

### Description

Full width numbers (`１２３４５６７８９`) are sometimes used by Japanese (or Chinese? etc like have kanjis), so I appreciate if there's option or basic support of it.

### PR DESCRIPTION
**Pull request by rikusen0335**
## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Refactoring
- [ ] Build or CI related changes
- [ ] Tests related changes
- [ ] Documentation content changes

## What is the current behaviour?
`maskitoNumberOptionsGenerator` does not accept japanese (also chinese?) full width numbers like: `０１２３４５６７８９`

Close #865

## What is the new behaviour?
It accepts full width numbers.

### CODE REVIEW

**Comment by rikusen0335** [comment]
Maybe I need to make new option to accept them, due to need of transformation of full width numbers...

**Comment by rikusen0335** [comment]
OOF

**Comment by rikusen0335** [comment]
Trying to run linters

**Comment by rikusen0335** [comment]
Lint done

File: projects/kit/src/lib/masks/number/processors/create-fullwidth-to-halfwidth-preprocessor.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
+import {MaskitoPreprocessor} from '@maskito/core';
+
+/**
+ * Convert full width numbers like １, ２ to half width numbers 1, 2
+ *
+ * 全角の１、２ののような数字を、半角の1、2のように直すプリプロセッサ
```
Comment (splincode) [inline projects/kit/src/lib/masks/number/processors/create-fullwidth-to-halfwidth-preprocessor.ts:6]: could you add e2e tests?

Comment (rikusen0335) [reply projects/kit/src/lib/masks/number/processors/create-fullwidth-to-halfwidth-preprocessor.ts:6]: Of course

Comment (rikusen0335) [reply projects/kit/src/lib/masks/number/processors/create-fullwidth-to-halfwidth-preprocessor.ts:8]: Adds the test, please confirm if it is completed as you desired, and if okay resolve conversation :)

Comment (splincode) [reply projects/kit/src/lib/masks/number/processors/create-fullwidth-to-halfwidth-preprocessor.ts:8]: thank you

**Comment by rikusen0335** [comment]
Can you run CIs again? Probably it is not code issue

**Comment by rikusen0335** [comment]
cspell says "fullwidth" is not known, then what would be great to say...?

**Comment by rikusen0335** [comment]
Okay, well, I added `fullwidth` and `halfwidth` to cspell configuration. Let me know if maskito team doesn't like it.

**Comment by splincode** [comment]
@rikusen0335 please revert your last commit

**Comment by rikusen0335** [comment]
Okay

**Comment by rikusen0335** [comment]
Reverted

**Comment by rikusen0335** [comment]
Auth error again, please run CI again

**Comment by splincode** [comment]
@rikusen0335 please check local failed spec file
image

**Comment by rikusen0335** [comment]
Fixed. Confirmed thousand separator and fullwidth replacement test works at least, and ran linter, prettier too

**Comment by rikusen0335** [comment]
local cspell still saying halfwidth is unknown word. Is this okay?

**Comment by splincode** [comment]
@rikusen0335 https://github.com/taiga-family/maskito/actions/runs/7437137144/job/20234500144?pr=864

passed

**Comment by rikusen0335** [comment]
Oh, I have to run `npm ci` instead of `npm run lint && npm run prettier && npm run cspell` xD

**Comment by splincode** [comment]
@rikusen0335 after the New Year holidays (after 9 Jan), that guys will review you

@nsbarsukov @waterplea @vladimirpotekhin

File: projects/kit/src/lib/masks/number/processors/create-fullwidth-to-halfwidth-preprocessor.ts (hunk 1 above)
Comment (nsbarsukov) [inline projects/kit/src/lib/masks/number/processors/create-fullwidth-to-halfwidth-preprocessor.ts:6]: Delete this line, please

English is enough

File: projects/kit/src/lib/masks/number/processors/create-fullwidth-to-halfwidth-preprocessor.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+        return {
+            elementState: {
+                selection,
+                value: value.replace(/[０-９]/g, s =>
+                    String.fromCharCode(s.charCodeAt(0) - 0xfee0),
+                ),
+            },
+            data: data.replace(/[０-９]/g, s =>
+                String.fromCharCode(s.charCodeAt(0) - 0xfee0),
+            ),
```
Comment (nsbarsukov) [inline projects/kit/src/lib/masks/number/processors/create-fullwidth-to-halfwidth-preprocessor.ts:21]: Let create function `toHalfWidthNumber` inside [`kit/src/lib/masks/number/utils`](https://github.com/taiga-family/maskito/tree/main/projects/kit/src/lib/masks/number/utils).

```ts
export function toHalfWidthNumber(fullWidthNumber: string): string {
    return fullWidthNumber.replace(/[０-９]/g, s =>
        String.fromCharCode(s.charCodeAt(0) - 0xfee0),
    )
}
```

and use it inside this preprocessor

```suggestion
                value: toHalfWidthNumber(value),
            },
            data: toHalfWidthNumber(data),
```

**Comment by rikusen0335** [comment]
Yup, thank you

File: projects/kit/src/lib/masks/number/processors/create-fullwidth-to-halfwidth-preprocessor.ts (hunk 2 above)
Comment (nsbarsukov) [reply projects/kit/src/lib/masks/number/processors/create-fullwidth-to-halfwidth-preprocessor.ts:21]: Could I also ask you to add some unit tests for this new utility `toHalfWidthNumber`-function (inside [`kit/src/lib/masks/number/utils/tests`](https://github.com/taiga-family/maskito/tree/main/projects/kit/src/lib/masks/number/utils/tests)) ?

```ts
import {stringifyNumberWithoutExp} from '../to-half-width-number';

describe('`toHalfWidthNumber` utility converts full width numbers to half width numbers', () => {
    it('０ => 0', () => {
        expect(toHalfWidthNumber('０')).toBe('0');
    });

    // [...] 
    // 2, 3, ..., 7, 8 

    it('９ => 9', () => {
        expect(toHalfWidthNumber('９')).toBe('9');
    });

    it('１２３４５６ => 123456', () => {
        expect(toHalfWidthNumber('１２３４５６')).toBe('123456');
    });

    it('１2３4５6 (full width + half width mix) => 123456', () => {
        expect(toHalfWidthNumber('１2３4５6 ')).toBe('123456');
    });
});
```

File: projects/kit/src/lib/masks/number/processors/index.ts [hunk 3]
```diff
@@ -1,3 +1,4 @@
+export * from './create-fullwidth-to-halfwidth-preprocessor';
```
Comment (nsbarsukov) [inline projects/kit/src/lib/masks/number/processors/index.ts:1]: Let's rename this file (to be similar to other same files with preprocessors)

```suggestion
export * from './fullwidth-to-halfwidth-preprocessor';
```

File: projects/kit/src/lib/constants/unicode-characters.ts [hunk 4]
```diff
@@ -35,3 +35,9 @@ export const CHAR_HYPHEN = '\u002D';
  * Can be used as `&minus;`. Don't confuse with {@link CHAR_HYPHEN}
  */
 export const CHAR_MINUS = '\u2212';
+
+/**
+ * {@link https://symbl.cc/en/30FC/ Katakana-Hiragana Prolonged Sound Mark}
+ * is used as prolonged sounds in Japanese.
+ */
+export const CHAR_JP_HYPHEN = '\u30FC';
```
Comment (nsbarsukov) [inline projects/kit/src/lib/constants/unicode-characters.ts:43]: Cool 🚀

**Review by nsbarsukov — COMMENTED** [review]
I didn't even know about this localization peculiarity. It is great to know it now, thanks!

Our team really appreciate your help ❤️

File: projects/kit/src/lib/masks/number/processors/create-fullwidth-to-halfwidth-preprocessor.ts (hunk 2 above)
Comment (rikusen0335) [reply projects/kit/src/lib/masks/number/processors/create-fullwidth-to-halfwidth-preprocessor.ts:21]: Yes sure

**Comment by rikusen0335** [comment]
Thank you for reviewing! I really appreciate you and team created really great stuff, also it's really clean lib I've ever seen!

Requested changes done.

File: projects/kit/src/lib/masks/number/utils/tests/to-half-width-number.spec.ts [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+        ['５', '5'],
+        ['６', '6'],
+        ['７', '7'],
+        ['８', '8'],
+        ['９', '9'],
+    ] as const;
+
+    tests.forEach(([fullWidthValue, halfWidthValue]) => {
+        it(`${fullWidthValue} => ${halfWidthValue}`, () => {
+            expect(toHalfWidthNumber('０')).toBe('0');
```
Comment (nsbarsukov) [inline projects/kit/src/lib/masks/number/utils/tests/to-half-width-number.spec.ts:19]:
```suggestion
            expect(toHalfWidthNumber(fullWidthValue)).toBe(halfWidthValue);
```

**Review by nsbarsukov — APPROVED** [review]

File: projects/kit/src/lib/masks/number/utils/tests/to-half-width-number.spec.ts (hunk 5 above)
Comment (rikusen0335) [reply projects/kit/src/lib/masks/number/utils/tests/to-half-width-number.spec.ts]: Forgot to fix that. Thank you!

**Review by splincode — APPROVED** [review]

**Comment by nsbarsukov** [comment]
This new feature will be release in Maskito `2.0.0` (the expected release date is the second half of January)

File: projects/kit/src/lib/masks/number/processors/fullwidth-to-halfwidth-preprocessor.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+export function createFullWidthToHalfWidthPreprocessor(): MaskitoPreprocessor {
+    return ({elementState, data}) => {
+        const {value, selection} = elementState;
+
+        return {
+            elementState: {
+                selection,
+                value: toHalfWidthNumber(value),
+            },
+            data: toHalfWidthNumber(data),
```
Comment (nsbarsukov) [inline projects/kit/src/lib/masks/number/processors/fullwidth-to-halfwidth-preprocessor.ts:17]: @rikusen0335 I have a small question to you again)

Should we use this same preprocessor for [`Time`](https://maskito.dev/kit/time), [`Date`](https://maskito.dev/kit/date), [`DateRange`](https://maskito.dev/kit/date-range), [`DateTime`](https://maskito.dev/kit/date-time) too ?

Are these built-in masks relevant for Japanese ?

Comment (rikusen0335) [reply projects/kit/src/lib/masks/number/processors/fullwidth-to-halfwidth-preprocessor.ts:17]: Ahhh probably yes, we can input full width numbers on them

Comment (nsbarsukov) [reply projects/kit/src/lib/masks/number/processors/fullwidth-to-halfwidth-preprocessor.ts:17]: @rikusen0335 Thank you!

I've created an issue for it
* https://github.com/taiga-family/maskito/issues/884

**Pull request merged**